File: hw1/task1.py
from pyspark.context import SparkContext
from pyspark.storagelevel import StorageLevel
import json
import sys

sc = SparkContext( 'local[*]', 'inf553_hw1_1' )  # initiate

input_file = sc.textFile( sys.argv[1] )  # readfile
data = input_file.map( lambda x: json.loads(x) ).map( lambda x: (x['review_id'], (x['user_id'], x['business_id'], x['useful'], x['stars'], len(x['text'])))).persist(StorageLevel(True, True, False, False)) # deal with json

museful = data.filter( lambda x: x[1][2] > 0 ).count()

mfivestar = data.filter( lambda x: x[1][3] == 5.0 ).count()

mlongestreview = data.map( lambda x: (x[1][4], 1) ).top(1)

muser = data.map(lambda x: (x[1][0], 1) ).reduceByKey(lambda x,y: x+y).sortByKey().persist(StorageLevel(True, True, False, False))

musernum = muser.count()

muserreview = muser.takeOrdered(20, lambda x: -x[1])

mbusiness = data.map(lambda x: ((x[1][1]), 1) ).reduceByKey(lambda x,y: x+y).sortByKey().persist(StorageLevel(True, True, False, False))

mbusinessnum = mbusiness.count()

mbusinessreview = mbusiness.takeOrdered(20, lambda x: -x[1])

ans_dict = {"n_review_useful": museful,
             "n_review_5_star": mfivestar,
             "n_characters": mlongestreview[0][0],
             "n_user": musernum,
             "top20_user": muserreview,
             "n_business": mbusinessnum,
             "top20_business": mbusinessreview}

ans_json = json.dumps(ans_dict)

ans_file=open( sys.argv[2], 'w')
ans_file.write(ans_json)
ans_file.close()

File: hw1/task1.scala
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.storage.StorageLevel
import org.json4s._
import org.json4s.jackson.JsonMethods.parse
import java.io.PrintWriter

object nan_zheng_task1 {
    def main(args: Array[String]): Unit = {

    val Array(srcFile, desFile) = args

    val sparkConf = new SparkConf().setAppName("inf553_hw1_scala").setMaster("local[*]")
    val sc = new SparkContext(sparkConf)

    val file = sc.textFile(srcFile)

    val data = file.map(x=>parse(x))
       .map(x=>((x\"review_id").values.asInstanceOf[String], (x\"user_id").values.asInstanceOf[String], (x\"business_id").values.asInstanceOf[String], (x\"useful").values.asInstanceOf[BigInt], (x\"stars").values.asInstanceOf[Double], (x\"text").values.asInstanceOf[String].length()))
       .persist(StorageLevel.MEMORY_AND_DISK_SER)

    val museful = data.filter( x => (x._4 > 0) ).count()

    val mfivestar = data.filter( x => x._5 == 5.0 ).count()

    val mlongestreview = data.map( x => (x._6, 1) ).top(1)

    val muser = data.map(x=>(x._2, 1)).reduceByKey(_+_).sortBy(x=>x._1).persist(StorageLevel.MEMORY_AND_DISK_SER)

    val musernum = muser.count()

    val muserreview = muser.sortBy(x=>x._2, false).take(20).map(x=>"["+"\""+x._1+"\", "+x._2+"]")

    val mbusiness = data.map(x=>(x._3, 1)).reduceByKey(_+_).sortBy(x=>x._1).persist(StorageLevel.MEMORY_AND_DISK_SER)

    val mbusinessnum = mbusiness.count()

    val mbusinessreview = mbusiness.sortBy(x=>x._2, false).take(20).map(x=>"[\""+x._1+"\", "+x._2+"]")


    var s:String = "{\"n_review_useful\": "+museful+", \"n_review_5_star\": "+mfivestar+", \"n_characters\": "+mlongestreview(0)._1+", \"n_user\": "+musernum+", \"top20_user\": ["
    for (x <- muserreview)  s += x
    s += "], \"n_business\": "+mbusinessnum+", \"top20_business\": ["
    for (x <- mbusinessreview)  s += x
    s += "]}"
    s = s.replace("][","], [")

    val writer = new PrintWriter(desFile)
    writer.println(s)
    writer.close()

    sc.stop()

  }


}